### PR TITLE
Simplify usage of Queues in nanny

### PR DESCRIPTION
**Pull request by fjetter**
This is a cleanup of the nanny around the usage of multiprocessing queues. It ensures that queues are only closed once and no exceptions are swallowed

### CODE REVIEW

File: distributed/nanny.py [hunk 1]
```diff
@@ -744,7 +741,7 @@ async def kill(self, timeout: float = 2, executor_wait: bool = True) -> None:
         if self.status == Status.stopping:
             await self.stopped.wait()
             return
-        assert self.status in (Status.starting, Status.running)
+        assert self.status in (Status.starting, Status.running, Status.failed)
```
Comment (fjetter) [inline distributed/nanny.py:744]: We actually hit this whenever a nanny failed. No idea if this actually has any implications or not. Not having this in actually deadlocked one of the unit tests at some point

File: distributed/nanny.py [hunk 2]
```diff
@@ -784,7 +779,7 @@ async def _wait_until_connected(self, uid):
                 continue
 
             if msg["uid"] != uid:  # ensure that we didn't cross queues
-                continue
+                raise RuntimeError("Encountered message from a different queue.")
```
Comment (fjetter) [inline distributed/nanny.py:782]: we're using a new queue for every process. how could these messages ever "cross"?

**Comment by graingert** [comment]
Would also be good to use socket.socketpair() instead of a Queue where the queue is only owned by two processes

**Review by hendrikmakait — APPROVED** [review]
LGTM! `test_lots_of_tasks` seems to be new, but doesn't look related to me.

**Pull request merged**